In [1]:
import numpy as np
import pandas as pd

In [2]:
#Initial parameters
input_vector=[1,3]
layer1_size = 2
weight_layer1 = np.array([[0.8, 0.5],[0.1, 0.2]])
layer2_size = 1
weight_layer2 = np.array([[0.2], [0.7]])
bias = 0
eta = 0.1
target = 0.95

In [3]:
class Perceptron:
    
    #initialize perceptron class
    def __init__(self, weights):
        self.weights = weights
        self.delta_weights = np.zeros_like(weights)
        self.activity = 0
        self.activation = 0
        self.delta = 0
        
        #Calculate the activity value
    def calc_activity(self, inputs):
        #print("Calc Activity first")
        #print("inputs: ", inputs)
        #print("weights: ", self.weights)
        self.activity = np.dot(inputs, self.weights)
        #print("Activity: ", self.activity)
        
    #Calculate the activation function given the activity value using sigmoid function
    def calc_activation(self, inputs):
        self.calc_activity(inputs)
        #print("Now activation: ")
        self.activation = 1/(1+np.exp(-1*self.activity))
        #print(self.activation)
        return self.activation
    
    def set_weights(self, new_weights):
        self.weights = new_weights


In [4]:
class Layer:
    
    #initialize layer class
    def __init__(self, layer_length, layer_weight, OutputLayerFlag= False):
        self.layer =  [Perceptron(weights=layer_weight[:, i]) for i in range(0,layer_length)]
        self.layer_length = layer_length 
        self.layer_weight = layer_weight
        self.OutputLayerFlag = OutputLayerFlag
        
    def get_layer_length(self):
        return self.layer_length
        
    def get_layer_output_vector(self, inputs):
        self.output_vector = [self.layer[i].calc_activation(inputs) for i in range(0, self.get_layer_length())]
        return np.array(self.output_vector)
    
    def get_error_vector(self, desired_output):
        self.littleE_vector = desired_output - np.array(self.output_vector)
        return self.littleE_vector
    
    def calc_layer_delta_weights(self, inputs, eta, prev_layer_delta=None, prev_layer_weights = None):
        if self.OutputLayerFlag:
            self.bigE = 0.5*np.array(self.littleE_vector)**2
            self.delta = self.littleE_vector*(1-np.array(self.output_vector))*self.output_vector
            self.delta_weights = np.array(self.layer_weight).T+ eta*self.delta*np.array(inputs)
        else:
            self.delta = (1-np.array(self.output_vector))*np.array(self.output_vector)*prev_layer_delta*np.array(prev_layer_weights).T
            self.delta_weights = self.layer_weight+eta*self.delta*inputs
        return self.delta_weights
    
    def update_layer_weights(self):
        self.layer_weight = self.delta_weights
        for i in range(0, self.get_layer_length()):
            self.layer[i].set_weights(self.delta_weights[i])
            print(self.layer[i].weights)   

In [5]:
Layer1=Layer(2,weight_layer1)
Layer2=Layer(1,weight_layer2, OutputLayerFlag = True)

In [6]:
Layer1.get_layer_output_vector(input_vector)

array([0.75026011, 0.75026011])

In [7]:
Layer2.get_layer_output_vector(Layer1.output_vector)

array([0.66267417])

In [8]:
Layer2.get_error_vector(target)

array([0.28732583])

In [9]:
Layer2.calc_layer_delta_weights(Layer1.output_vector, eta)

array([[0.20481877, 0.70481877]])

In [10]:
Layer1.calc_layer_delta_weights(input_vector, eta, Layer2.delta, Layer2.layer_weight)

array([[0.80024069, 0.50252722],
       [0.10024069, 0.20252722]])

In [11]:
Layer1.delta

array([[0.00240688, 0.00842407]])

In [12]:
#Layer2.update_layer_weights()
#Layer1.update_layer_weights()

Question 1: What is the initial activation function value of Node 1? 0.7503

In [13]:
Layer1.layer[0].activation

0.7502601055951177

Question 2: What is the initial activation value of Node 2? 0.7503

In [14]:
Layer1.layer[1].activation

0.7502601055951177

Quesion 3: What is the initial activation value of Node 3?  0.6627

In [15]:
Layer2.layer[0].activation

0.6626741680915721

Question 4: Using your previously calculated values, what is the updated weight for the link connecting node 1 with node 3?  0.2048

In [16]:
Layer2.calc_layer_delta_weights(Layer1.output_vector, eta)[:,0]

array([0.20481877])

Question 5: Using the data you've calculated, what is the updated weight for the link connecting Node 2 with Node 3? 0.7048

In [17]:
Layer2.calc_layer_delta_weights(Layer1.output_vector, eta)[:,1]

array([0.70481877])

Question 6: For the updated weight from input 1 to node 2, by how much does the weight increase or decrease?  0.0008

In [18]:
Layer1.delta[0,1]

0.008424073202270744